File: ParameterPassingMechanisms/CallByValueResult.java
/* Ryan Hirscher
 * Dr. Oudshoorn
 * CSC4510
 * Assignment 2 Call By Value Result Java
 *
 * Java is sctrictly built to support pass by value. Addresses are not able to be retrieved for variables.
 *
 * Java does not support Call by Value-Result directly. However, it can be implmeneted to accomplish the same feature through defining your own class objects. These objects are referenced through pointers under the hood, so changes made to the formal parameters will be reflected back to the actuals. This utilizes aliasing to accomplish call by value result.
 *
 * Actual value changes are reflected in main.
*/

//Callee and defining object class CallByVal
class CallByVal {
   int X, Y;

   //Method for assignment
   CallByVal(int x, int y)
   {
      X=x;
      Y=y;
   }

   //Method to add
   static void addFive(CallByVal obj)
   {
         System.out.println("Formals: X=" + obj.X + " Y=" + obj.Y);

         //Copying/adding back at the time of return
         obj.X+=5;
         obj.Y+=5;
         return;
   }
}

//Caller
class CallByValueResult {
   public static void main(String[] args)
   {
      //Declaring object type CallByVal named obj
      //New instance and storing 5 as X and 12 as Y
      CallByVal obj = new CallByVal(5, 12);

      System.out.println("Call By Value Result (class object) to addFive function.");
      System.out.println("Actuals Pre-add: X=" + obj.X + " Y=" + obj.Y);

      //Calling
      obj.addFive(obj);

      System.out.println("Actuals Post-add: X=" + obj.X + " Y=" + obj.Y);
   }
}

File: ParameterPassingMechanisms/CallByValueResult.cpp
/* Ryan Hirscher
 * Dr. Oudshoorn
 * CSC4510
 * Assignment 2 Call by Value-result C++
 *
 * This program attempts to demonstrate C++ pass by value-result mechanism by examining the scope and memory locations. It is important to note that C++ does not support value-result, but we can mimic its feature with its native Pass by Reference mechanism.
 *
 * Actual and formal variables shared the same address and values were swapped. The swap retained outside of the function.
*/

#include<iostream>
using namespace std;
void swap (int &, int &);
int main()
{
   int x=10;
   int y=999;

   //What is the address of the actual values? - Same memory locations in the function
   cout << "Call by Value-Result to swap function" << endl;
   cout << "Pre-swap actual X: " << x << endl;
   cout << "X memory address: " << &x << endl;
   cout << "Pre-swap actual Y: " << y << endl;
   cout << "Y memory address: " << &y << endl << endl;
   
   //Passing the actual
   swap(x, y);

   //Did the memory location of the actual values change? - They did not change memory locations
   //Did the values change from the function? - They swapped outside of the scope as well
   cout << "Post-swap actual X: " << x << endl;
   cout << "X memory address: " << &x << endl;
   cout << "Post-swap actual Y: " << y << endl;
   cout << "Y memory address: " << &y << endl << endl;
   
   return 0;
}

void swap (int &x, int &y)
{
   int temp=0;
   temp = x;
   x = y;
   y = temp;

   //Do x and y retain the same memory locations? - They had the same memory locations
   //Did the values swap? - They swapped locally and retained the swap outside the function
   cout << "In-swap formal X: " << x << endl;
   cout << "X memory address: " << &x << endl;
   cout << "In-swap formal Y: " << y << endl;
   cout << "Y memory address: " << &y << endl << endl;

   return;
}
